### ISSUE TITLE
phase-2 reco CPU hotspot: HGCDoublet::checkCompatibilityAndTag

### ISSUE THREAD

**Issue by slava77** [cms-sw/cmssw#30957]
(pointed out by @dpiparo )

HGCDoublet::checkCompatibilityAndTag is found to consume 2.8% of CPU time in workflow 23434.21 (step3); of which almost 1/4 is in ```int& std::vector<int >::emplace_back```
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/CMSSW_11_2_X_2020-07-21-2300-sign1106.376a0992c5d.23434.21.step3.50.pp/70
(this is as of CMSSW_11_2_X_2020-07-21-2300, before #30815 was merged)

https://github.com/cms-sw/cmssw/blob/cc4483924d15af7843f759ee3231a37dab4222d9/RecoHGCal/TICL/plugins/HGCDoublet.cc#L3

Estimate gain: 0.5%, depends

Things to improve: reserve? 

For tests, the following can provide a minimal setup
```
runTheMatrix.py -l 23434.21  -w upgrade -i all  --command=" -n 50 --customise Validation/Performance/TimeMemoryInfo.py --lazy_download" -t 4 >& pu.log &
```
A speedup visible with just Timing service (or FastTiming if preferred) can be enough to show speedup. Igprof analysis for update code is welcome as well.

@rovere @felicepantaleo

**Comment by slava77** [comment]
assign reconstruction,upgrade

**Comment by felicepantaleo** [comment]
@slava77 , I will look at it starting from next week.

**Comment by rovere** [comment]
@slava, do we have a measurement/profile after #30815 has been merged?

**Comment by slava77** [comment]
>  Maybe I've found it: [link](https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_2_0_pre4/75)

this should work; but my test in the PR description was done on a different machine. So, the numbers may not be comparable.

**Comment by perrotta** [comment]
> @Slava, do we have a measurement/profile after #30815 has been merged?
> Maybe I've found it: [link](https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_2_0_pre4/75)

Exact.
#30815 was included since pre3, already (in the same page)

**Comment by rovere** [comment]
> > Maybe I've found it: [link](https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_2_0_pre4/75)
> 
> this should work; but my test in the PR description was done on a different machine. So, the numbers may not be comparable.

Yes, indeed, I was looking at the percentage measurements that show a little improvement since.

**Comment by slava77** [comment]
there are also per-IB reports available now from https://cmssdt.cern.ch/SDT/html/cmssdt-ib/#/ib/CMSSW_11_2_X
fewer events, but much more regular
e.g. from CMSSW_11_2_X_2020-09-01-2300
https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_11_2_X_2020-09-01-2300/slc7_amd64_gcc820/profiling
https://cmssdt.cern.ch/SDT/cgi-bin/igprof-navigator/CMSSW_11_2_X_2020-09-01-2300/slc7_amd64_gcc820/profiling/23434.21/igprofCPU_step3

**Comment by felicepantaleo** [comment]
@slava77 
by applying: https://github.com/cms-sw/cmssw/compare/master...felicepantaleo:reserve_emplace_back

we go from 
http://fpantale.web.cern.ch/fpantale/cgi-bin/igprof-navigator/TICL_PU200_CMSSW_11_2_0_pre5/54

to 
http://fpantale.web.cern.ch/fpantale/cgi-bin/igprof-navigator/TICL_PU200_CMSSW_11_2_0_pre5_PR/50

I suppose the time is spent in constructing objects rather than allocate memory. What do you think?

**Comment by slava77** [comment]
@felicepantaleo 
thank you for sharing the test results.
I'm curious if the shrink_to_fit is the call that increases the cost so much.

Note thought that https://github.com/cms-sw/cmssw/compare/master...felicepantaleo:reserve_emplace_back changes `allDoublets_` in HGCGraphT (which is a `vector<HGCDoublet`), while this github issue description is about `vector<int` inside HGCDoublet.
This, apparently is about `tagAsOuterNeighbor` and `tagAsInnerNeighbor` calls; and the suggestion to reserve would apply to `outerNeighbors_` and/or `innerNeighbors_`

**Comment by slava77** [comment]
+1

this kind of fixed itself via #31906 (TICL v3 update in November 2020)
`int& std::vector<int >::emplace_back` cost is  0.05% in CMSSW_11_3_0_pre2  https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_3_0_pre2/298

the change can be seen by comparing 11_2_0_pre8 https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_2_0_pre8/73 with 11_2_0_pre9 https://jiwoong.web.cern.ch/jiwoong/cgi-bin/igprof-navigator/igprofCPU_CMSSW_11_2_0_pre9/317 ( #31906 was merged in pre9)
